enhancements/network/external-lb-vips.md

```diff
@@ -0,0 +1,268 @@
+the VIPs used on the cluster control plane.
+
+### Goals
+
+Allow to replace the internal loadbalancer by an external one and support this scenario when the customer
+prefers to maintain it themselves. Later we provide more details on what is expected from the customer to do.
+
+### Non-Goals
+
+* Making this scenario by default at some point. The internal loadbalancer for on-premise platforms will
+  remain the HAproxy/Keepalived/CoreDNS stack.
+* Change the default behavior of the VSphere UPI workflow.
+* Support the migration from the internal to the external loadbalancer (and vice-versa). This will be done in
+  a separate enhancement proposal in the future.
+* Support the ability to change the IP addresses of the VIPs on the loadbalancer. This will be done in a
+  separate enhancement proposal in the future.
+  will have to choose a scenario.
+
+## Proposal
+
+We will make some change in `install-config.yaml` so customers can configure the
+deployment to use their own loadbalancer if they want.
+Then the components which interact with networking will have to check this
+new field and handle this scenario appropriately.
+
+### Workflow Description
+
+1. The customer will configure their external loadbalancer(s), which includes configuring the VIPs
+   for API and Ingress and also configure the loadbalancer tool that is being used (e.g. HAproxy, F5, etc)
+   Note that we don't support Static IPs yet, so the deployment will have to configure the LB backends
+   with all the IPs that are available in the subnet of the control-plane machines.
+2. The customer will write the `install-config.yaml` file and set the new `loadBalancerType` platform specific field to `UserManaged`, and
+   will set the fields for `apiVIP` and `ingressVIP`. Since this is will be TechPreview at first,
+   customers will have to set `featureSet: TechPreviewNoUpgrade`.
+3. In the installer, if the platform does support external loadbalancer, some validation will be run
+   to check that the VIPs can already be resolved (the validation already exists for vsphere, we need to
+    make it re-usable). The new field for `LoadBalancerType` will be set in `PlatformStatus`.
+4. In machine-config-operator, if LB type is external, the HAproxy and Keepalived will not be
+   deployed.
+
+### API Extensions
+
+A new field will be added to the `PlatformStatus` of the infrastructure object to tell
+the kind of the loadbalancer.
+
+```golang
+// PlatformLoadBalancerType defines the type of load balancer used by the cluster.
+type PlatformLoadBalancerType string
+const (
+    // LoadBalancerTypeUserManaged is a loadbalancer that is managed outside of the cluster by the customer.
+    LoadBalancerTypeUserManaged PlatformLoadBalancerType = "UserManaged"
+
+    // LoadBalancerTypeOpenShiftManaged is a loadbalancer that is managed by the OpenShift cluster.
+    LoadBalancerTypeOpenShiftManaged PlatformLoadBalancerType = "OpenShiftManaged"
```

Comment:
Is there a third possibility here, where the load balancer is configured by the installer (through Terraform) but not otherwise managed by OpenShift? Which category would that fall into?

Maybe that's not an issue because we're restricting the option to on-prem platforms that otherwise use keepalived.

Reply:
the new API proposal in the next patchset will address it.